### PR TITLE
Physical S3 backend implementation

### PR DESCRIPTION
**Pull request by jstremick**
Adds a new 's3' physical backend to store vault data in an S3 bucket. 

Notes:
* The bucket must exist before vault can use it and we'll verify that it can be read.
* Credentials will sought from a chain of sources: backend config -> environment -> credential file -> EC2 IAM role
* ~~Recursive delete test added. Looking at the other backends, it looks like if a key is deleted that any child keys should also be deleted.~~

No HA support as the eventually consistent nature of S3 means acquiring locks via bucket isn't feasible. Maybe SimpleDB or another synchronous service could provide that ability?

I updated godeps.json ~~should I also commit the libraries it updated?~~  and aws-sdk-go.

### CODE REVIEW

**Comment by jstremick** [comment]
Implements #120

**Comment by jstremick** [comment]
Looks like I've broken tests for a few other physical backends when testing recursive deletes. Let me know what the expected behavior should be and I'll update.

The S3Backend tests fail without credentials. Should I do anything about that?

**Comment by certifiedloud** [comment]
+1, I'd love to see this feature working in Vault.

File: physical/physical_test.go [hunk 1]
```diff
@@ -189,6 +189,21 @@ func testBackend_ListPrefix(t *testing.T, b Backend) {
 	if keys[0] != "baz" {
 		t.Fatalf("bad: %v", keys)
 	}
+
+	// Delete should recursively remove paths
+	err = b.Delete("foo")
```
Comment (armon) [inline physical/physical_test.go]: This should not be the case actually. This is invalid. If "foo" is deleted, only that key should be. The directories under should not be affected.

File: physical/s3.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	s := &S3Backend{
+		client: s3conn,
+		bucket: bucket,
+	}
+	return s, nil
+}
+
+// Put is used to insert or update an entry
+func (s *S3Backend) Put(entry *Entry) error {
+	s.l.Lock()
```
Comment (armon) [inline physical/s3.go]: This internal locking is not necessary, there is no need to serialize the access to S3.

File: physical/s3.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	return ent, nil
+}
+
+// Delete is used to permanently delete an entry
+func (s *S3Backend) Delete(key string) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	defer metrics.MeasureSince([]string{"s3", "delete"}, time.Now())
+
+	listResp, err := s.client.ListObjects(&s3.ListObjectsInput{
```
Comment (armon) [inline physical/s3.go]: Delete will name only a single object, so you should be able to do a `DeleteObject` against the exact named key.

File: physical/s3_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"os"
+	"testing"
+	"time"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/service/s3"
+)
+
+func TestS3Backend(t *testing.T) {
+
```
Comment (armon) [inline physical/s3_test.go]: Something like this: https://github.com/hashicorp/vault/blob/master/physical/consul_test.go#L13 should be done to skip the test if we do not have the credentials provided.

**Comment by armon** [comment]
@jstremick I've left some feedback, but this is looking really good!

**Comment by armon** [comment]
@jstremick Awesome, thanks!

**Comment by jstremick** [comment]
@armon Welcome!

**Pull request merged**